### PR TITLE
Improvement: Build without -disablewallet

### PR DESCRIPTION
**Pull request by ch4ot1c**
### Working towards a successful build without `-disablewallet`:

- Prefer `uint64_t` over `size_t` (See https://github.com/zcash/zcash/pull/3236)
- Add `pzcashParams` and related
- Logging modernized (Uses `BCLog` and `gArgs`) 
- Replace `CBitcoinAddress` with `CTxDestination` and related

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -1161,7 +1161,7 @@ AX_CHECK_COMPILE_FLAG([-fwrapv],[CXXFLAGS="$CXXFLAGS -fwrapv"])
 AX_CHECK_COMPILE_FLAG([-fno-strict-aliasing],[CXXFLAGS="$CXXFLAGS -fno-strict-aliasing"])
 AX_CHECK_COMPILE_FLAG([-Wno-builtin-declaration-mismatch],[CXXFLAGS="$CXXFLAGS -Wno-builtin-declaration-mismatch"],,[[$CXXFLAG_WERROR]])
 
-LIBZCASH_LIBS="-lgmp -lgmpxx $BOOST_SYSTEM_LIB -lcrypto -lsodium $RUST_LIBS"
+LIBZCASH_LIBS="-lgomp -lgmp -lgmpxx $BOOST_SYSTEM_LIB -lcrypto -lsodium $RUST_LIBS"
```
Comment (jc23424) [inline configure.ac]: What is this needed for?

File: src/init.cpp [hunk 2]
```diff
@@ -289,6 +293,10 @@ void Shutdown()
     GetMainSignals().UnregisterWithMempoolSignals(mempool);
     g_wallet_init_interface.Close();
     globalVerifyHandle.reset();
+
+    delete pzcashParams;
```
Comment (jc23424) [inline src/init.cpp]: if you declare as a `std::unique_ptr` above you don't have to manually invoke the destructor here

File: src/init.h [hunk 3]
```diff
@@ -21,6 +23,8 @@ namespace boost
 class thread_group;
 } // namespace boost
 
+extern ZCJoinSplit* pzcashParams;
```
Comment (jc23424) [inline src/init.h]: better to make this a unique_ptr so it gets automatically cleaned up

File: src/util.cpp [hunk 4]
```diff
@@ -750,9 +750,7 @@ static boost::filesystem::path ZC_GetBaseParamsDir()
         pathRet = fs::path(pszHome);
 #ifdef MAC_OSX
     // Mac
-    pathRet /= "Library/Application Support";
-    TryCreateDirectory(pathRet);
-    return pathRet / "ZcashParams";
+    pathRet / "Library/Application Support/ZcashParams";
```
Comment (jc23424) [inline src/util.cpp]: this line has no effect? you may want to assign this value somewhere

File: src/wallet/asyncrpcoperation_mergetoaddress.cpp [hunk 5]
```diff
@@ -23,16 +25,15 @@
...
 #include <chrono>
 #include <iostream>
 #include <string>
 #include <thread>
 
-#include <wallet/paymentdisclosuredb.h>
-
 using namespace libzcash;
-
+#if 0
```
Comment (jc23424) [inline src/wallet/asyncrpcoperation_mergetoaddress.cpp]: better to just remove this file from compilation right now instead of ifdefing a bunch of it out?
then mergetoaddress can be a separate pr?

File: src/wallet/asyncrpcoperation_sendmany.h [hunk 6]
```diff
@@ -21,7 +21,7 @@
 
 // Default transaction fee if caller does not specify one.
 #define ASYNC_RPC_OPERATION_DEFAULT_MINERS_FEE   10000
-
+#if 0
```
Comment (jc23424) [inline src/wallet/asyncrpcoperation_sendmany.h]: same as mergetoaddress

File: src/wallet/asyncrpcoperation_shieldcoinbase.cpp [hunk 7]
```diff
@@ -108,9 +106,9 @@ void AsyncRPCOperation_shieldcoinbase::main() {
 
 #ifdef ENABLE_MINING
   #ifdef ENABLE_WALLET
-    GenerateBitcoins(false, NULL, 0);
```
Comment (jc23424) [inline src/wallet/asyncrpcoperation_shieldcoinbase.cpp]: probably can just remove this entirely

File: src/wallet/asyncrpcoperation_shieldcoinbase.cpp [hunk 8]
```diff
@@ -30,11 +30,9 @@
 
 #include <wallet/asyncrpcoperation_shieldcoinbase.h>
 
-#include <wallet/paymentdisclosure.h>
-#include <wallet/paymentdisclosuredb.h>
 
 using namespace libzcash;
-
+#if 0
```
Comment (jc23424) [inline src/wallet/asyncrpcoperation_shieldcoinbase.cpp]: same

File: configure.ac (hunk 1 above)
Comment (ch4ot1c) [reply configure.ac]:
```
snark/libsnark.a(profiling.o): In function `libsnark::enter_block(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool)':
profiling.cpp:(.text+0x14e7): undefined reference to `GOMP_critical_start'
profiling.cpp:(.text+0x1558): undefined reference to `GOMP_critical_end'
snark/libsnark.a(profiling.o): In function `libsnark::leave_block(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool)':
profiling.cpp:(.text+0x1cc5): undefined reference to `GOMP_critical_start'
profiling.cpp:(.text+0x1ed6): undefined reference to `GOMP_critical_end'
collect2: error: ld returned 1 exit status
```

Comment (ch4ot1c) [reply configure.ac]: When building on Ubuntu

**Comment by jc23424** [comment]
utACK - in the future you might find `git rebase -i` to be useful to help you squash and reorder your commits after you go through a round of reviews etc...

File: src/util.cpp (hunk 4 above)
Comment (donshin) [inline src/util.cpp]: @ch4ot1c for some reason I'm not seeing this change in master after your merge- was it removed intentionally?

Comment (ch4ot1c) [reply src/util.cpp]: Yes; was fixed to Zcash's implementation:
```
    // Mac
    pathRet /= "Library/Application Support";
    TryCreateDirectory(pathRet);
    return pathRet / "ZcashParams";
```

Comment (ch4ot1c) [reply src/util.cpp]: But certainly should instead be https://github.com/BTCPrivate/BTCP-Rebase/pull/36, good catch @donshin.
